### ISSUE TITLE
cubecl × CUDA graph: interleaved cubecl ops between replays corrupt subsequent captures/replays

### ISSUE THREAD

**Issue by Harry-Zhou** [tracel-ai/cubecl#1612]
2026-09-03 after an 11-discriminator isolation session on the NexusMind
graph-decode driver (Ornith-1.5-35B offloaded serving, ~2361-node captured
decode steps).

## Environment

- cubecl 0.10 (vendored with visibility patches: `CudaServer.streams` pub,
  `resolve_device_ptr`, `get_cu_stream`, `ComputeClient::with_server`), via
  burn-cubecl 0.21 / burn-backend 0.21
- CUDA 12.2, RTX 3090, driver-managed stream-ordered allocations
  (`cuMemAllocAsync` on the device default mempool)
- Repro: `cargo test -p nexusmind-inference --features w4-cuda,cuda-graph
  --test ornith_dense_proxy_bench -- --ignored graph_capture_repro --nocapture`
  (~5 GB synthetic checkpoint, ~5 min)

## Symptom

A CUDA graph captured from cubecl kernel launches (topology: ~2100 kernel
nodes + ~250 D2D memcpy nodes, **zero** MEM_ALLOC/MEM_FREE/EVENT nodes — all
activations served from cubecl's internal free lists) replays fine in
isolation, but a *subsequent* capture (or sometimes the next replay of the
same graph) faults inside the graph with `CUDA_ERROR_ILLEGAL_ADDRESS`.
compute-sanitizer pins the faulting access to a cubecl `#[cube]` kernel
reading gigabytes past one of its array arguments (a garbage-computed index
or a stale baked pointer):

```
Invalid __global__ read of size 4 bytes
  at moe_expert_batched_kernel ... by thread (64,0,0) in block (1,0,0)
  Address 0x20f7ee5804 is out of bounds — 13,873,825,285 bytes after the
  nearest allocation (24,697,856 bytes)
  Host Frame: cuGraphLaunch → replay_stream_graph
```

## Discriminator matrix (all same process/session)

| # | Interleaved between capture/replay ops | Result |
|---|---|---|
| E | nothing (3-4 consecutive replays of one graph) | ✅ works |
| G | one host→device staging allocation + flush + sync + drop | ✅ works |
| H | 4 replays, then a **second capture**, then ITS replay | ✅ works |
| — | first replay that READS graph outputs (miss-flag `into_data`), then any second replay (same or recaptured graph) | ❌ illegal address |
| — | replay + GDN state D2D backup/restore raw `cuMemcpyAsync` + pos-buffer writes | ❌ (before output-read isolation) |
| — | full sequential model forward between replays | ❌ |

Additional constants ruled out: destroying the previous graph (`F` — leaking
it changes nothing), the default mempool release threshold (set to
`UINT64_MAX` — no effect), host-pointer memcpys (none exist — every memcpy
node is D2D `(0,0)`), autotune shape changes (shapes identical), GDN state
tensor replacement (poison-flagged separately, absent in the crashing runs).

The topology histogram differs slightly between captures (~±20 nodes) —
cubecl's scheduler emits a non-deterministic op mix for the same forward.

## Working hypotheses (for the cubecl maintainers)

1. `Tensor::into_data()` on a buffer whose allocation participated in a
   capture (graph-managed free-list block) advances some scheduler state
   (binding cursors / drop-queue fences) that the NEXT capture then bakes
   incorrectly — e.g. a stale device pointer or a lost dependency edge.
2. The `PendingDropQueue` background thread frees a block asynchronously
   (fence-gated) that a later capture has already baked into kernel args;
   the pool holds the VA mapped (threshold MAX) but the graph's private
   state diverges from cubecl's memory-management bookkeeping.
3. Some capture-adjacent path in `MemoryManagement`/`align_streams` records
   cross-stream event ops only when bindings cross streams; the capture
   drops them (we see zero EVENT nodes even though multi-stream scheduling
   is active), so replay executes unordered launches — a racing
   `slot_resolve` write vs the batched-expert kernel read would produce
   exactly the observed garbage-slot OOB. Why the first replays stay correct
   would then be scheduling luck.

## Resolution (2026-09-03 follow-up session)

A clean single-variable staged harness (T0-T6, one op + one replay per stage
on the proxy) isolated the trigger precisely:

| Stage | Op between two replays | Result |
|---|---|---|
| T0 | nothing (pure consecutive replays) | ✅ |
| T6 | raw `cuMemcpyAsync` DtoH read of the baked logits | ✅ |
| T3 | `into_data` read of a FRESH (post-capture) tensor | ✅ |
| T4-raw | raw `cuMemcpyAsync` DtoD write into the baked x_static | ✅ |
| T5b | one sequential model forward with NaN guards OFF | ✅ |
| T1 | cubecl `into_data` read of the baked logits | ❌ ILLEGAL_ADDRESS |
| T2 | cubecl `into_data` read of a baked miss-flag | ❌ |
| T4-cubecl | a cubecl kernel (`slab_write_scales`) writing a baked buffer | ❌ |
| T5 | one sequential forward with guards ON (their `into_data`s trip the class) | ❌ |

**Conclusion**: the poison class is *any cubecl read/allocator/kernel operation
that touches a buffer whose address the captured graph baked* — executed
*between* replays. Raw driver calls (`cuMemcpyDtoHAsync`/`cuMemcpyDtoDAsync`/
`cuMemcpyHtoDAsync`) on the same stream are completely harmless. The
mechanism is a cubecl-internal state interaction (its read path allocates
pinned staging + creates `cuEvent` fences + resolves resource cursors; its
write kernels allocate staging from the same free list that holds the graph's
freed intermediates) that the graph's replay then mis-executes against.

## What NexusMind does (shipped)

The graph-decode driver now runs all inter-replay bookkeeping through raw
driver calls (D2H reads for logits/miss-flags/positions, D2D/H2D writes for
x-content/slot-maps/pos-restore/GDN rollback) and sequential fallback steps
run with the NaN guards disabled (their reads are the poison; the forward's
own reads touch only fresh tensors and are safe). Captured once, the graph
then survives the full adaptive loop (proxy: 10 consecutive replays GREEN,
zero misses; real-weight Ornith-35B E2E validating token-level equivalence
is in flight). The `cubecl_graph_slab` tests still pass unchanged.

## FINAL RESOLUTION (2026-09-03, third session) — root cause chain fully closed

The staged harness was rebuilt with a CLEAN pre-loop state (the earlier
harness's own mutations had confounded it), adding per-layer GDN NaN
reports on the live state AND the rollback backups. This exposed TWO
stacked root causes:

### Root cause 1 — pool-backup aliasing (the second-cycle NaN)

The driver's rollback BACKUP tensors were allocated from cubecl's pool
AFTER the capture. Pool slices handed out post-capture can alias the
graph's freed capture-time intermediates. The REPLAY then overwrites
them; the rollback restores the overwritten (NaN-containing) garbage
into the live GDN state; the sequential re-step NaNs.

Diagnostic evidence (proxy, forced-miss cycle 0):
```
[step0/backup]            GDN state: all-clean
[step0/post-replay]       GDN state: all-clean
[step0/BACKUP-after-replay] conv NaN layers [1,2,3,4]   ← the replay corrupted the backups
rollback fidelity (delta L0-2): true                     ← the D2D copies are byte-faithful
[step0/restored]          GDN NaN: conv layers [1,2,3,4] ← garbage restored into the state
```

Fix: all replay-crossing buffers (the rollback backups) are RAW
`cuMemAlloc` driver memory — outside the pool entirely.

### Root cause 2 — graph intermediates on reusable pool slices (the ILLEGAL_ADDRESS class)

The capture-time intermediates were served from cubecl's pool FREE
LISTS (no driver allocation → no MEM_ALLOC nodes — see the topology
histograms). Those slices return to the free lists at capture end; ANY
post-capture pool allocation (guards' reads, sequential forwards'
activations, staging) can land on them; the next replay then reads
clobbered intermediates → garbage indices (the slab kernel's GB-past
reads) or boundary faults (the reduce kernel's +1 read).

Fix (vendored cubecl patch — `memory_management::graph_capture`):
- During a stream capture, pool pages are allocated SYNCHRONOUSLY
  (`cuMemAlloc`, NOT `cuMemAllocAsync` — async would become a graph
  MEM_ALLOC node, and such graphs fail their SECOND launch with
  CUDA_ERROR_INVALID_VALUE, empirically).
- Capture-time pages are FROZEN with the capture generation: served
  only while THAT capture is active, never re-served afterwards, never
  deallocated (a new capture bumps the generation, retiring older
  frozen pages).
- `MemoryManagement::reserve` skips the persistent pool during capture.

With both fixes, the FULL adaptive loop passes on the proxy: miss →
rollback → guards-ON sequential re-step → re-arm → replay-again across
MULTIPLE cycles, plus clean HIT replays — and the historical T1/T2/
T4-cubecl/T5 poison stages (guards-on forwards, cubecl reads/writes of
baked buffers between replays) are all neutralized by the quarantine.

### Upstreamable ask

The pool-layer graph-capture mode is the general fix other
capture-based users would need: "allocations made while a graph
capture is active must not be served from, or returned to, the general
free lists". Our implementation is in vendor/cubecl-runtime
(sliced_pool.rs + memory_manage.rs + the graph_capture flag module) and
vendor/cubecl-cuda (sync alloc during capture).
